### PR TITLE
Replay recording extension until results screen transition

### LINKED ISSUE #22493: Replay recording should continue to record until the results screen transition
I agree that this is behaviour we want to keep around.

### Discussed in https://github.com/ppy/osu/discussions/22481

<div type='discussions-op-text'>

<sup>Originally posted by **monochrome22** February  1, 2023</sup>
One thing that always bothered me in lazer is that replays end immediately after pressing the last object. In stable, it will continue recording for a few seconds, until the player goes to the results screen. Often during the few seconds after the map ends, players do movements with their cursors or even use smoke. 

Here are some examples: 
(rrtyui) [https://youtu.be/17PP3XB1IzM?t=131](https://youtu.be/17PP3XB1IzM?t=131) (skip to 2:11)
(Bikko) [https://youtu.be/xQmuYsXmBH4?t=135](https://youtu.be/xQmuYsXmBH4?t=135) (skip to 2:15)
(Cookiezi) [https://youtu.be/86pK4kJ1v_g?t=55](https://youtu.be/86pK4kJ1v_g?t=55) (skip to 0:55)
(aetrna) [https://youtu.be/hJmXbwF55rM?t=195](https://youtu.be/hJmXbwF55rM?t=195) (skip to 3:15)

While this is how it works in lazer:
[https://youtu.be/H2FeD34Pfew?t=85](https://youtu.be/H2FeD34Pfew?t=85) (skip to 1:25)

I feel that the movements that players do at the end of the replay add charm and personality to both the player and the play. I always thought it's cool how rrtyui flicks his cursor to the top of his screen, and seeing Bikko do that smooth curve after spinner adds a sense of calmness to his plays. Having the replay end immediately takes these away and makes it more dull.

I don't have suggestions on exactly how it should be implemented. Just have it record for a few extra seconds or until players goes to results screen.
</div>

### PR DESCRIPTION
**Pull request by Terochi**
Fixes https://github.com/ppy/osu/issues/22493

Works similarly to stable.

Replay ends when:
- player presses the Escape (`GlobalAction.Back`) key
- player skips storyboard ending
- one second (`Player.RESULTS_DISPLAY_DELAY`) after last `HitObject`
- one second (`Player.RESULTS_DISPLAY_DELAY`) after storyboard ending

### CODE REVIEW

File: osu.Game/Screens/Play/Player.cs [hunk 1]
```diff
@@ -800,7 +768,13 @@ private void progressToResults(bool withDelay)
 
             resultsDisplayDelegate = new ScheduledDelegate(() =>
             {
-                if (prepareScoreForDisplayTask?.IsCompleted != true)
+                if (prepareScoreForDisplayTask == null)
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs:771]: Isn't this going to run multiple times? This delegate is looping.

Comment (Terochi) [reply osu.Game/Screens/Play/Player.cs:771]: Potentially yes in cases where the the `beginScoreImport` returns on the if check. I don't necessarily see it as an issue because it is essentially behaving as before, and the `Scheduler` might stop working soon once the screen exits (not sure about this behavior). If desired for the `canShowResults ` check can be used at the beginning of `progressToResults` as well.

Comment (peppy) [reply osu.Game/Screens/Play/Player.cs:771]: I dunno, I can't follow this logic. Like why is this only run when the task is null? There's too much missing detail which makes it hard to follow the changes.

**Comment by Terochi** [comment]
I will try to make the code clearer to read

File: osu.Game/Screens/Play/Player.cs [hunk 2]
```diff
@@ -770,7 +771,11 @@ private void progressToResults(bool withDelay)
             {
                 if (prepareScoreForDisplayTask == null)
                 {
-                    beginScoreImport();
+                    // Try importing score since the task hasn't been invoked yet.
+                    if (!beginScoreImport())
+                        // If the task hasn't started, the score will never be imported.
+                        resultsDisplayDelegate?.Cancel();
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs:777]: What does this even mean? The comment makes me even more confused.

File: osu.Game/Screens/Play/Player.cs [hunk 3]
```diff
@@ -790,17 +795,25 @@ private void progressToResults(bool withDelay)
             Scheduler.Add(resultsDisplayDelegate);
         }
 
-        private void beginScoreImport()
+        /// <summary>
+        /// Ends replay recording and runs <see cref="prepareAndImportScore"/> only when results can be shown
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs:799]: This xmldoc shares nothing in common with the method name. Summaries in xmldoc should *expand* on the method name, not describe two different processes.

**Comment by peppy** [comment]
I've made some further changes to hopefully make the whole process easier to follow.

I do think it's also feasible to replace the `resultsDisplayDelegate` logic with an `async` invocation instead (with one initially delayed schedule), but don't want to do that in this PR as there are other locations in `Player` which cancel that process, and I want to avoid breaking them.

I'm pretty happy with this now, but will need a double-check to make sure it feels good @ppy/team-client.

**Review by peppy — APPROVED** [review]

File: osu.Game/Screens/Play/Player.cs [hunk 4]
```diff
@@ -816,6 +798,51 @@ private void progressToResults(bool withDelay)
...
+        /// <summary>
+        /// Asynchronously run score preparation operations (database import, online submission etc.).
+        /// </summary>
+        /// <param name="forceImport">Whether the score should be imported even if non-passing (or the current configuration doesn't allow for it).</param>
+        /// <returns>The final score.</returns>
+        [ItemCanBeNull]
+        private Task<ScoreInfo> prepareAndImportScoreAsync(bool forceImport = false)
+        {
+            // Ensure we are not writing to the replay any more, as we are about to consume and store the score.
+            DrawableRuleset.SetRecordTarget(null);
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/Player.cs:810]: Should I be worried that this is changing `Drawable` states via calling `.Expire()` internally, yet it runs on a TPL thread when the button's clicked from the results screen?

image

Comment (peppy) [reply osu.Game/Screens/Play/Player.cs:810]: It's pretty dodge when you look at it that way, but the `null` path at most [calls `Expire()` on a drawable](https://github.com/ppy/osu/blob/7bc8908ca9c026fed1d831eb6e58df7624a8d614/osu.Game/Rulesets/UI/RulesetInputManager.cs#L45) so i guess it's mostly safe...

**Review by smoogipoo — APPROVED** [review]
